File: app/src/main/java/com/heinika/designpattern/factory/ShapeFactory.java
package com.heinika.designpattern.factory;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;

public class ShapeFactory {
    private static final String shape_PACKAGE = "com.heinika.designpattern.factorypattern.";
    public static Shape getshape(String className){
        className = shape_PACKAGE + className;
        try {
            Class c = Class.forName(className);
            Constructor constructor = c.getConstructor();
            return (Shape) constructor.newInstance();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: app/src/main/java/com/heinika/designpattern/kotlin/factory/ShapeFactory.kt
package com.heinika.designpattern.kotlin.factory

import java.lang.Exception

class ShapeFactory {
    fun getShape(shapeName: String): Shape? {
        try {
            val c = Class.forName("com.heinika.designpattern.kotlin.factory.$shapeName")
            return c.getDeclaredConstructor().newInstance() as Shape
        }catch (e : Exception){
            return null
        }

    }
}